sunpy/coordinates/spice.py

```diff
@@ -0,0 +1,275 @@
+system is being increasingly used by space missions to describe the locations of
+spacecraft and the time-varying orientations of reference frames.
+While the `~spiceypy.spiceypy` package provides a Python interface for
+performing SPICE computations, its API is very different from that of
+`~astropy.coordinates.SkyCoord`.
+
+This module "wraps" `~spiceypy.spiceypy` functionality so that relevant SPICE
+computations can be accessed using the `~astropy.coordinates.SkyCoord` API.
+When loading a set of kernels, a frame class and corresponding transformations
+are created for each SPICE frame.  One can also query the location of a body
+as computed via SPICE.
+
+See :ref:`sphx_glr_generated_gallery_units_and_coordinates_spice.py` for an
+example of how to use this module.
+
+Notes
+-----
+* 2D coordinates can be transformed only if the to/from frames have the same
+  SPICE body ID as their centers.
+* Transformations of velocities are not yet supported.
+* SPICE frame names are rendered as uppercase, except for plus/minus characters,
+  which are replaced with lowercase ``'p'``/``'n'`` characters.
+"""
+
+import numpy as np
+
+try:
+    import spiceypy
+except ImportError:
+    raise ImportError("This module requires the optional dependency `spiceypy`.")
+
+import astropy.units as u
+from astropy.coordinates import ICRS, ConvertError, SkyCoord, frame_transform_graph
+from astropy.coordinates.representation import CartesianRepresentation
+from astropy.coordinates.transformations import FunctionTransformWithFiniteDifference
+from astropy.time import Time
+
+from sunpy import log
+from sunpy.coordinates import SunPyBaseCoordinateFrame
+from sunpy.time import parse_time
+from sunpy.time.time import _variables_for_parse_time_docstring
+from sunpy.util.decorators import add_common_docstring
+
+__all__ = ['get_body', 'initialize']
+
+
+# Note that this epoch is very slightly different from the typical definition of J2000.0 (in TT)
+_ET_REF_EPOCH = Time('J2000', scale='tdb')
+
+_CLASS_TYPES = {2: 'PCK', 3: 'CK', 4: 'TK', 5: 'dynamic', 6: 'switch'}
+
+
+# Registry of the generated frame classes and center classes
+_frame_registry = {}
+_center_registry = {'SOLAR SYSTEM BARYCENTER': ICRS}
+
+
+# Defined for future use
+class _SpiceBaseCoordinateFrame(SunPyBaseCoordinateFrame):
+    pass
+
+
+def _convert_to_et(time):
+    return (time - _ET_REF_EPOCH).to_value('s')
+
+
+def _astropy_frame_name(spice_frame_name):
+    # Replace plus/minus characters in the SPICE frame name with lowercase 'p'/'n'
+    return f"spice_{spice_frame_name.translate(str.maketrans('+-', 'pn'))}"
+
+
+def _astropy_center_name(spice_center_id):
+    # Use the center ID directly, changing a negative sign to 'n'
+    return f"_spice_center_{str(spice_center_id).replace('-', 'n')}"
+
+
+def _is_2d(data):
+    return data.norm().unit is u.one and u.allclose(data.norm(), 1*u.one)
+
+
+def _install_center_by_id(center_id):
```

Comment:
Can you explain to me (and maybe in a docstring) a little more about what a "center" is here. I am a little worried that there's a lot of stuff which is :magic_wand: here if you aren't that familiar with spice.

Reply:
I added some developer notes